### PR TITLE
Add out of bound check for Leader Key sequence array

### LINKED ISSUE #5762: Multiple random keypresses after KC_LEAD freezes keyboard.
I'm not sure I'd characterize this as a bug or just misuse of the LEADER feature. If I hit KC_LEAD then mash several keys quickly my keyboard completely freezes. I realize this isn't the intent of the LEADER feature, but it's not an implausible scenario.

I have a layer that locks my keyboard so anyone (like say a 2 year old) can mash on the keyboard without registering keypresses. It's unlocked with KC_LEAD ...  I've found an easy way to avoid the freeze issue, but didn't know if this is an expected behavior that just should be avoided.

## System Information
Massdrop ALT on Windows
QMK release 0.6.341

### PR DESCRIPTION
**Pull request by drashna**
Leader keys is limited to 5 keypresses, even the array for is sets it to 5, but at no time does it check to make sure that it's hit the end of the array before adding more key sequences.... which appears to crash the controller. 

## Description

This adds a sanity check to make sure that once you've hit 5 keys, it doesn't add more.  However, it forcibly ends the lead sequence, and may not process it correctly. 

## Types of Changes

- [x] Core
- [x] Bugfix

## Issues Fixed or Closed by This PR

* Fixes #5762

## Checklist

- [x] My code follows the code style of this project.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/process_keycode/process_leader.c [hunk 1]
```diff
@@ -58,8 +58,13 @@ bool process_leader(uint16_t keycode, keyrecord_t *record) {
           keycode = keycode & 0xFF;
         }
 #endif // LEADER_KEY_STRICT_KEY_PROCESSING
-        leader_sequence[leader_sequence_size] = keycode;
-        leader_sequence_size++;
+        if (leader_sequence_size < 5) {
```
Comment (yanfali) [inline quantum/process_keycode/process_leader.c]: maybe `sizeof(leader_sequence)`

**Comment by yanfali** [comment]
I also don't like the assignment to init memory. Just use `memset(leader_sequence, 0, sizeof(leader_sequence))`, it should generate better code and be less maintenance if the array changes size.

**Comment by drashna** [comment]
I'm not sure that would work, since qk_lead_start sets the values for all five values in the array, but sets them to 0. 

Also, I'm not rally sure how memset works...

**Comment by yanfali** [comment]
http://man7.org/linux/man-pages/man3/memset.3.html

**Comment by drashna** [comment]
@yanfali You mean like this:

```diff
diff --git a/quantum/process_keycode/process_leader.c b/quantum/process_keycode/process_leader.c
index 30ce2a023..43bd4410b 100644
--- a/quantum/process_keycode/process_leader.c
+++ b/quantum/process_keycode/process_leader.c
@@ -33,19 +33,13 @@ bool leading = false;
 uint16_t leader_time = 0;
 
 uint16_t leader_sequence[5] = {0, 0, 0, 0, 0};
-uint8_t leader_sequence_size = 0;
 
 void qk_leader_start(void) {
   if (leading) { return; }
   leader_start();
   leading = true;
   leader_time = timer_read();
-  leader_sequence_size = 0;
-  leader_sequence[0] = 0;
-  leader_sequence[1] = 0;
-  leader_sequence[2] = 0;
-  leader_sequence[3] = 0;
-  leader_sequence[4] = 0;
+  memset(leader_sequence, 0, sizeof(leader_sequence));
 }
 
 bool process_leader(uint16_t keycode, keyrecord_t *record) {
@@ -58,9 +52,8 @@ bool process_leader(uint16_t keycode, keyrecord_t *record) {
           keycode = keycode & 0xFF;
         }
 #endif // LEADER_KEY_STRICT_KEY_PROCESSING
-        if (leader_sequence_size < 5) {
-          leader_sequence[leader_sequence_size] = keycode;
-          leader_sequence_size++;
+        if (sizeof(leader_sequence) < 5) {
+          leader_sequence[sizeof(leader_sequence)] = keycode;
         } else {
           leading = false;
           leader_end();
diff --git a/quantum/process_keycode/process_leader.h b/quantum/process_keycode/process_leader.h
index 15bccc3f6..0bdab7c0f 100644
--- a/quantum/process_keycode/process_leader.h
+++ b/quantum/process_keycode/process_leader.h
@@ -32,7 +32,7 @@ void qk_leader_start(void);
 #define SEQ_FOUR_KEYS(key1, key2, key3, key4) if (leader_sequence[0] == (key1) && leader_sequence[1] == (key2) && leader_sequence[2] == (key3) && leader_sequence[3] == (key4) && leader_sequence[4] == 0)
 #define SEQ_FIVE_KEYS(key1, key2, key3, key4, key5) if (leader_sequence[0] == (key1) && leader_sequence[1] == (key2) && leader_sequence[2] == (key3) && leader_sequence[3] == (key4) && leader_sequence[4] == (key5))
 
-#define LEADER_EXTERNS() extern bool leading; extern uint16_t leader_time; extern uint16_t leader_sequence[5]; extern uint8_t leader_sequence_size
+#define LEADER_EXTERNS() extern bool leading; extern uint16_t leader_time; extern uint16_t leader_sequence[5]
 #define LEADER_DICTIONARY() if (leading && timer_elapsed(leader_time) > LEADER_TIMEOUT)
 
 #endif
```

**Comment by yanfali** [comment]
p much. It should compile, it might need a header include.

File: quantum/process_keycode/process_leader.c [hunk 2]
```diff
@@ -33,19 +33,13 @@ bool leading = false;
...
 
 uint16_t leader_sequence[5] = {0, 0, 0, 0, 0};
-uint8_t leader_sequence_size = 0;
 
 void qk_leader_start(void) {
   if (leading) { return; }
   leader_start();
   leading = true;
   leader_time = timer_read();
-  leader_sequence_size = 0;
```
Comment (yanfali) [inline quantum/process_keycode/process_leader.c]: you have to reset this value. Put it back please.

File: quantum/process_keycode/process_leader.c [hunk 3]
```diff
@@ -58,9 +52,8 @@ bool process_leader(uint16_t keycode, keyrecord_t *record) {
           keycode = keycode & 0xFF;
         }
 #endif // LEADER_KEY_STRICT_KEY_PROCESSING
-        if (leader_sequence_size < 5) {
-          leader_sequence[leader_sequence_size] = keycode;
-          leader_sequence_size++;
```
Comment (yanfali) [inline quantum/process_keycode/process_leader.c]: you can't remove this without adding it to the line above. e.g. leader_sequence++

File: quantum/process_keycode/process_leader.h [hunk 4]
```diff
@@ -32,7 +32,7 @@ void qk_leader_start(void);
 #define SEQ_FOUR_KEYS(key1, key2, key3, key4) if (leader_sequence[0] == (key1) && leader_sequence[1] == (key2) && leader_sequence[2] == (key3) && leader_sequence[3] == (key4) && leader_sequence[4] == 0)
 #define SEQ_FIVE_KEYS(key1, key2, key3, key4, key5) if (leader_sequence[0] == (key1) && leader_sequence[1] == (key2) && leader_sequence[2] == (key3) && leader_sequence[3] == (key4) && leader_sequence[4] == (key5))
 
-#define LEADER_EXTERNS() extern bool leading; extern uint16_t leader_time; extern uint16_t leader_sequence[5]; extern uint8_t leader_sequence_size
```
Comment (yanfali) [inline quantum/process_keycode/process_leader.h]: why is this a 16bit value now? It should just be an 8, it only goes up to 5.

File: quantum/process_keycode/process_leader.c [hunk 5]
```diff
@@ -58,9 +52,8 @@ bool process_leader(uint16_t keycode, keyrecord_t *record) {
...
+        if (sizeof(leader_sequence) < 5) {
+          leader_sequence[sizeof(leader_sequence)] = keycode;
```
Comment (yanfali) [inline quantum/process_keycode/process_leader.c]: OK, my bad. this should be leader_sequence_size not sizeof(leader_sequence). Move that to the 5 above.

**Comment by drashna** [comment]
Yeah, going to revert most of this, I think.

File: quantum/process_keycode/process_leader.h (hunk 4 above)
Comment (drashna) [reply quantum/process_keycode/process_leader.h]: it wasn't.  `leader_sequence` is what stores the keycodes.  `leader_sequence_size` is what counts the number of entries in the sequence.

Comment (yanfali) [reply quantum/process_keycode/process_leader.h]: the counter is separate from the storage.

**Comment by drashna** [comment]
You mean more like this?

**Comment by shinmai** [comment]
Won't compile for me without including `string.h`, but when included fixes #5762 on my Planck and Preonic at least.

**Comment by jpetermans** [comment]
Same. Won't compile until string.h is included, but with that, try as I might, mashing as many keys as I can after KC_LEADER does not lock up the keyboard.

**Comment by drashna** [comment]
That's ... odd. It compiles just fine for me....

Where are you including string.h?

**Comment by jpetermans** [comment]
This compiles fine for me and fixes the issue on my ALT.

**Comment by drashna** [comment]
Awesome!

**Comment by jpetermans** [comment]
I know there's always a lot going on with PR's. Checking in when this one will get merged. @drashna's changes work fine for my setup.

**Comment by drashna** [comment]
@jpetermans could you got to the "files" tab, and "review" the PR, and approve it?  

That will help to get it pushed through. :)

File: quantum/process_keycode/process_leader.c (hunk 1 above)
Comment (drashna) [reply quantum/process_keycode/process_leader.c]: you mean like:
```suggestion
        if (leader_sequence_size < sizeof(leader_sequence)) {
```

**Comment by patrickwelker** [comment]
I just pulled from `qmk:master` and tried to compile, but got an error. This might be related to the recent changes:

```
Compiling: keyboards/ergodox_ez/keymaps/patrick/keymap.c                                            [OK]
Compiling: quantum/process_keycode/process_leader.c                                                quantum/process_keycode/process_leader.c: In function 'qk_leader_start':
quantum/process_keycode/process_leader.c:47:3: error: implicit declaration of function 'memset' [-Werror=implicit-function-declaration]
   47 |   memset(leader_sequence, 0, sizeof(leader_sequence));
      |   ^~~~~~
quantum/process_keycode/process_leader.c:47:3: error: incompatible implicit declaration of built-in function 'memset' [-Werror]
quantum/process_keycode/process_leader.c:20:1: note: include '<string.h>' or provide a declaration of 'memset'
   19 | #include "process_leader.h"
  +++ |+#include <string.h>
   20 | #ifdef __arm__
cc1: all warnings being treated as errors
 [ERRORS]
 |
 |
 |
make[1]: *** [.build/obj_ergodox_ez_patrick/quantum/process_keycode/process_leader.o] Error 1
make: *** [ergodox_ez:patrick] Error 1
Make finished with errors
```

**Comment by fauxpark** [comment]
Yeah, string.h is required for memset() on AVR as well as ARM... that ifdef shouldn't be guarding it 😄

**Comment by drashna** [comment]
Whoops.  I swear it compiled on AVR just fine.   #6347 opened to fix that.

**Comment by drashna** [comment]
@patrickwelker pull again, and this should be fixed, I hope.

**Comment by patrickwelker** [comment]
@drashna Yep. Everything compiles without an error again without manually correcting `process_leader.c`. Thanks.

**Comment by drashna** [comment]
Glad to hear it!  Thanks for catching that!

**Pull request merged**